File: front-end/flutter/livee/lib/data/core/api_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:livee/data/core/session_storage_service.dart';

class ApiClient {
  static const String _apiBase = 'https://main-server-ekgr.onrender.com/api/v1';

  static final ApiClient _instance = ApiClient._internal();
  factory ApiClient() => _instance;
  ApiClient._internal();

  Future<Map<String, String>> _getAuthHeaders() async {
    final token = SessionStorageService.read('liveeToken');
    final headers = <String, String>{
      'Content-Type': 'application/json',
    };
    if (token != null) {
      headers['Authorization'] = 'Bearer $token';
    }
    return headers;
  }

  Future<http.Response> get(String path) async {
    final uri = Uri.parse('$_apiBase$path');
    final headers = await _getAuthHeaders();
    return http.get(uri, headers: headers);
  }

  Future<http.Response> post(String path, {Map<String, dynamic>? body}) async {
    final uri = Uri.parse('$_apiBase$path');
    final headers = await _getAuthHeaders();
    return http.post(uri, headers: headers, body: jsonEncode(body));
  }

  Future<http.Response> put(String path, {Map<String, dynamic>? body}) async {
    final uri = Uri.parse('$_apiBase$path');
    final headers = await _getAuthHeaders();
    return http.put(uri, headers: headers, body: jsonEncode(body));
  }

  Future<http.Response> delete(String path) async {
    final uri = Uri.parse('$_apiBase$path');
    final headers = await _getAuthHeaders();
    return http.delete(uri, headers: headers);
  }
}

File: front-end/react/lib/apiClient.ts
import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';

const API_BASE = 'https://main-server-ekgr.onrender.com/api/v1';

class ApiClient {
    private axiosInstance: AxiosInstance;

    constructor() {
        this.axiosInstance = axios.create({
            baseURL: API_BASE,
            headers: {
                'Content-Type': 'application/json',
            },
        });

        this.axiosInstance.interceptors.request.use(async (config) => {
            // TODO: 로컬 스토리지에서 토큰을 가져와 헤더에 추가하는 로직 구현
            const token = localStorage.getItem('liveeToken');
            if (token) {
                config.headers.Authorization = `Bearer ${token}`;
            }
            return config;
        });
    }

    public async get<T>(path: string, config?: AxiosRequestConfig): Promise<T> {
        const response = await this.axiosInstance.get<T>(path, config);
        return response.data;
    }

    public async post<T>(path: string, body: any, config?: AxiosRequestConfig): Promise<T> {
        const response = await this.axiosInstance.post<T>(path, body, config);
        return response.data;
    }

    public async put<T>(path: string, body: any, config?: AxiosRequestConfig): Promise<T> {
        const response = await this.axiosInstance.put<T>(path, body, config);
        return response.data;
    }

    public async delete<T>(path: string, config?: AxiosRequestConfig): Promise<T> {
        const response = await this.axiosInstance.delete<T>(path, config);
        return response.data;
    }
}

export const apiClient = new ApiClient();